stl/inc/xstring

```diff
@@ -3764,7 +3822,8 @@ public:
         return size() == 0;
     }
 
-    size_type copy(_Out_writes_(_Count) _Elem* const _Ptr, size_type _Count, const size_type _Off = 0) const {
+    _CONSTEXPR20_MEMOP size_type copy(
+        _Out_writes_(_Count) _Elem* const _Ptr, size_type _Count, const size_type _Off = 0) const {

```

Comment:
This change, modifying `basic_string`, should be reverted. [P1032R1](https://wg21.link/P1032R1) doesn't modify `basic_string` at all. (I was initially confused, because its sections 11 and 12 depict two `copy` changes - but they're changing the `basic_string_view` class definition in `[string.view.template]` and the `basic_string_view` member function definition in `[string.view.ops]`. Bonus confusion: the latter has a section name of `"String operations"`.)

It's #43 [P0980R1](https://wg21.link/P0980R1) `constexpr` `std::string`, which @mnatsuhara is currently investigating, that will change `basic_string::copy`. (I find it amusing that C++20 had so many `constexpr` papers, leading to double bonus confusion here - misreading P1032R1 as changing `basic_string::copy` won't be error-corrected by checking the Working Paper because it was changed there by another paper! :joy_cat:)

As an aside, which is now moot, this would have to be `_CONSTEXPR20`, not `_CONSTEXPR20_MEMOP`, since this calls `_Traits::copy` to do work, and the user might have provided a custom `constexpr` traits class.

Reply:
@StephanTLavavej Can I get some clarification as to why a potentially user-provided `constexpr` traits class requires `_CONSTEXPR20` instead of `_CONSTEXPR20_MEMOP`?  (I imagine this will also apply to `_CONSTEXPR20_DYNAMIC_ALLOC` in [P0980R1](https://wg21.link/P0980R1)).

Reply:
@mnatsuhara Sure! `_CONSTEXPR20` says "`constexpr` in C++20 (and later), `inline` in C++17 (and earlier)". `_CONSTEXPR20_MEMOP` is a refinement, saying "C++20 says this should be `constexpr`, but we specifically need support for `constexpr` compiler builtins that only Clang provides at the moment". So currently in this PR, the `char_traits` functions that are directly calling compiler builtins for Clang, and otherwise calling C functions, need to be marked `_CONSTEXPR20_MEMOP` to reflect this implementation limitation.

However, this doesn't apply transitively to `basic_string_view::copy`. C++20 marks `basic_string_view::copy` as `constexpr`, and it calls `_Traits::copy` to do work. Our `char_traits::copy` is marked `_CONSTEXPR20_MEMOP` because of the use of builtins - but the `basic_string_view` could have been given a user-defined character traits class, and the user might have implemented a `constexpr` `copy` with the plain Core Language. In that case, marking `basic_string_view::copy` as `_CONSTEXPR20_MEMOP` would be unnecessarily limiting - it would be non-`constexpr` for MSVC in C++20 mode, even though the user provided a `constexpr`-capable traits class and there is no use of the compiler builtin (or C `memcpy`). So `basic_string_view::copy` should be marked as `_CONSTEXPR20`, which will then make it `constexpr` in C++20 mode **if** the traits class has `constexpr` `copy` (e.g. `char_traits` for Clang, or user-defined traits that implemented `constexpr`).

> (I imagine this will also apply to _CONSTEXPR20_DYNAMIC_ALLOC in P0980R1).

I believe you're correct - the same "customization" logic applies, with allocators instead of traits. Even if an implementation doesn't support `constexpr` `new` and `delete`, as long as our `allocator_traits` machinery is `_CONSTEXPR20`, a user should be able to implement a `constexpr`-compatible pool allocator (although maybe not due to the lack of `reinterpret_cast`?).

In general, I see little potential harm from applying `_CONSTEXPR20_MEOW` in a targeted, non-transitive manner (i.e. only in the places that directly need additional compiler support). If customization through a template parameter (allocator/traits/etc.) is possible, then either it actually matters for users (like `copy`) so `_CONSTEXPR20` is correct, or it turns out not to be possible to customize in a `constexpr` fashion without compiler support, but the compiler will never realize that so it won't emit "no `constexpr` ever possible, this is ill-formed no-diagnostic-required and I can emit a diagnostic here". The only scenario I can envision seeing a compiler complain about is where we mark an internal function as `_CONSTEXPR20_MEOW`, we mark an outer/public layer as `_CONSTEXPR20`, there are no sources of customization along the way, and the compiler notices that we have an always-`constexpr` (in C++20) function calling an always-non-`constexpr` function (without `MEOW` support), and it emits the not-actually-required diagnostic - in which case we can just go fix it up.

Reply:
Resolving as I'm reverting the `basic_string` change tracked here - but if I can further clarify the `constexpr` situation, please let me know.